Src/EB/AMReX_EB_chkpt_file.cpp

```diff
@@ -0,0 +1,353 @@
+
+    std::string line, word;
+
+    std::getline(is, line);
+
+    int nlevs;
+    is >> nlevs;
+    gotoNextLine(is);
+    AMREX_ASSERT(nlevs == 1);
+
+    std::getline(is, line);
+    {
+        std::istringstream lis(line);
+        int i = 0;
+        while (lis >> word) {
+            prob_lo[i++] = std::stod(word);
+        }
+    }
+
+    std::getline(is, line);
+    {
+        std::istringstream lis(line);
+        int i = 0;
+        while (lis >> word) {
+            prob_hi[i++] = std::stod(word);
+        }
+    }
+
+    for (int idim = 0; idim < AMREX_SPACEDIM; ++idim) {
+        AMREX_ALWAYS_ASSERT_WITH_MESSAGE(Math::abs(prob_lo[idim] - geom.ProbLo()[idim]) < std::numeric_limits<Real>::epsilon(),
+                                         "EB2::ChkptFile cannot read from a different problem domain");
+        AMREX_ALWAYS_ASSERT_WITH_MESSAGE(Math::abs(prob_hi[idim] - geom.ProbHi()[idim]) < std::numeric_limits<Real>::epsilon(),
+                                         "EB2::ChkptFile cannot read from a different problem domain");
+    }
+
+    IntVect ngrow_chkptfile;
+    std::getline(is, line);
+    {
+        std::istringstream lis(line);
+        int i = 0;
+        while (lis >> word) {
+            ngrow_chkptfile[i++] = std::stoi(word);
+        }
+    }
+    AMREX_ALWAYS_ASSERT_WITH_MESSAGE(ngrow_chkptfile == ngrow_finest, "EB2::ChkptFile cannot read from different ngrow");
+
+    bool edf_chkptfile;
+    is >> edf_chkptfile;
+    gotoNextLine(is);
+    AMREX_ALWAYS_ASSERT_WITH_MESSAGE(extend_domain_face == edf_chkptfile,
+                                     "EB2::ChkptFile cannot read from different extend_domain_face");
+
+    int mgs_chkptfile;
+    is >> mgs_chkptfile;
+    gotoNextLine(is);
+    AMREX_ALWAYS_ASSERT_WITH_MESSAGE(max_grid_size == mgs_chkptfile,
+                                     "EB2::ChkptFile cannot read from different max_grid_size");
+
+    BoxArray cut_grids_ba, covered_grids_ba;
+
+    if (amrex::Verbose()) amrex::Print() << "Loading cut_grids\n";
+    cut_grids_ba.readFrom(is);
+    gotoNextLine(is);
+
+    if (is.peek() != EOF) {
+        if (amrex::Verbose()) amrex::Print() << "Loading covered_grids\n";
+        covered_grids_ba.readFrom(is);
+        gotoNextLine(is);
+    }
+
+    BoxList cut_bl;
+    for (int nb = 0; nb < cut_grids_ba.size(); nb++) {
+        Box b(cut_grids_ba[nb]);
+        cut_bl.push_back(b);
+    }
+
+    BoxList covered_bl;
+    for (int nb = 0; nb < covered_grids_ba.size(); nb++) {
+        Box b(covered_grids_ba[nb]);
+        covered_bl.push_back(b);
+    }
+
+    cut_grids.define(cut_bl);
+    dmap.define(cut_grids, ParallelDescriptor::NProcs());
+
+    covered_grids.define(covered_bl);
+
+    // volfrac
+    {
+        if (amrex::Verbose()) amrex::Print() << "  Loading " << m_volfrac_name << std::endl;
+
+        volfrac.define(cut_grids, dmap, 1, ng_gfab);
+        volfrac.setVal(1.); // regular value
```

Comment:
I don't think we need to call setVal on any of the MultiFabs that will be read from the checkpoint file.

Reply:
Thanks for the comment. The reason I was having a `setVal` with the regular value was to handle the ghost values from invalid regions because they fall into fully regular grids. Since we have special handling for covered cells and cells outside the domain, I was thinking this is necessary for regular cells. The other option is to combine this logic into `set_invalid_ghost_data_covered`. Please let me know if this makes more sense and I can do that.

Reply:
My point is whatever is set in valfrac.setVal() will be overwritten by VisMF::Read that touches every cell.

Reply:
Ah I see your point. I incorrectly thought that `setVal` will set all cells including the invalid ghost cells to the regular value and VisMF::Read will only touch the valid cells. But looks like `VisMF::Read` will overwrite everything as you said. I think I need to handle this in a separate method like `set_invalid_ghost_data_regular`.

Reply:
I might have missed something here, but I don't understand why you need to have something like `set_invalid_ghost_data_regular`.  `VisMF::Wrtie` writes out everything including ghost cells.  `VisMF::Read` restores everything in the memory.  So whatever was in the ghost cells before the checkpoint file was written will be completely restored after `VisMF::Read`.  If we didn't need `set_invalid_ghost_regular` before, we don't need it now.

Reply:
When I try writing using `VisMF::Write` on a geometry which has cut, covered and regular grids and then read using `VisMF::Read`, I find that the ghost cells are NaNs on a cut grid that has a covered or a regular grid as neighbor. My understanding was that this is because we only store cut grids and optimize out the regular and covered grids. So the regular and covered grids would fall into the "invalid" region of the MultiFab and hence those ghost cells are not read. Is this not the correct behavior? thanks.

Reply:
`VisMF::Read` reads everything including ghost cells.  If there are NaNs after `VisMF::Read`, there were there already when `VisMF::Write` was called.  Do you have a test case demonstrating what you mentioned above?

Reply:
I see. If that is the case I guess I shouldn't be needing `set_invalid_ghost_data_extended` or `set_invalid_ghost_data_covered` either, since they should get read from the checkpoint file. Is this right?

I tried this [case](https://ccse.lbl.gov/pub/RegressionTesting1/MFIX-Exa/2022-08-26/BENCH05-Size0008-coveredgrid.inputs.rt.covered_grid) adding `eb2.max_grid_size=10`
1. use [eb-restart](https://mfix.netl.doe.gov/gitlab/rangarad/mfix/-/tree/eb_restart) branch of mfix from my fork.
2. first run the case with `mfix.write_geom_chk=1`
3. then run the case with `mfix.geometry="chkptfile"`

I find that the `volfrac` on cell `(-18,-18,-12)` for example seems to give nan after reading from the checkpoint file. After the call to `set_invalid_ghost_data_extended`, it becomes a valid value.

Reply:
you are right though that the nan was already there when calling `VisMF::write`. I will revisit this tomorrow. Looks like when I switched from the first method of
```
MultiFab mf(The_Pinned_Arena());
VisMF::Read(mf, prefix);
volfrac.ParallelCopy(mf, 0, 0, 1, 0, ng_gfab, geom.periodicity());
```
to
```
VisMF::Read(levelset, prefix);
```
it has made things more straightforward. I don't think I need the `setVal`, `set_invalid_ghost_data_extended` or `set_invalid_ghost_data_covered` any more. i will test this and update the PR again.

Thanks a lot for your help with this!